src/NServiceBus.Core/Recoverability/RecoverabilityExecutor.cs

```diff
@@ -7,42 +7,31 @@
 
     class RecoverabilityExecutor
     {
-        public RecoverabilityExecutor(IRecoverabilityPolicy recoverabilityPolicy, IEventAggregator eventAggregator, DelayedRetryExecutor delayedRetryExecutor, MoveToErrorsExecutor moveToErrorsExecutor, bool transactionsOn)
+        public RecoverabilityExecutor(bool raiseRecoverabilityNotifications, Func<RecoverabilityConfig, ErrorContext, RecoverabilityAction> recoverabilityPolicy, RecoverabilityConfig configuration, IEventAggregator eventAggregator, DelayedRetryExecutor delayedRetryExecutor, MoveToErrorsExecutor moveToErrorsExecutor)
         {
+            this.configuration = configuration;
             this.recoverabilityPolicy = recoverabilityPolicy;
             this.eventAggregator = eventAggregator;
             this.delayedRetryExecutor = delayedRetryExecutor;
             this.moveToErrorsExecutor = moveToErrorsExecutor;
-            this.transactionsOn = transactionsOn;
 
-            raiseNotifications = recoverabilityPolicy.RaiseRecoverabilityNotifications;
-            delayedRetriesEnabled = delayedRetryExecutor != null;
+            raiseNotifications = raiseRecoverabilityNotifications;
+            delayedRetriesCapabilityAvailable = delayedRetryExecutor != null;
         }
 
         public Task<ErrorHandleResult> Invoke(ErrorContext errorContext)
         {
-            if (transactionsOn == false || errorContext.Exception is MessageDeserializationException)
-            {
-                return MoveToError(errorContext);
-            }
-
-            return PerformRecoverabilityAction(errorContext);
-        }
-
-        Task<ErrorHandleResult> PerformRecoverabilityAction(ErrorContext errorContext)
-        {
-            var recoveryAction = recoverabilityPolicy.Invoke(errorContext);
+            var recoveryAction = recoverabilityPolicy.Invoke(configuration, errorContext);
 
             if (recoveryAction is ImmediateRetry)
             {
                 return RaiseImmediateRetryNotifications(errorContext);
             }
 
-            // When we can't do delayed retries then just fall through to error.
-            if (recoveryAction is DelayedRetry && !delayedRetriesEnabled)
+            // When we can't do delayed retries, a policy customization probably didn't honor MaxNumberOfRetries for DelayedRetries
+            if (recoveryAction is DelayedRetry && !delayedRetriesCapabilityAvailable)
```

Comment:
Same as above. Maybe we should go the defensive path here and fallback to `MoveToErrors` instead of throwing?